typing_extensions/src_py3/typing_extensions.py

```diff
@@ -2166,3 +2169,226 @@ class TypeAlias(metaclass=_TypeAliasMeta, _root=True):
         It's invalid when used anywhere except as in the example above.
         """
         __slots__ = ()
+
+
+# Python 3.10+ has PEP 612
+if hasattr(typing, 'ParamSpec'):
+    ParamSpec = typing.ParamSpec
+else:
+    # Inherits from list as a workaround for Callable checks in Python < 3.9.2.
+    class ParamSpec(list):
+        """Parameter specification variable.
+
+        Usage::
+
+           P = ParamSpec('P')
+
+        Parameter specification variables exist primarily for the benefit of static
+        type checkers.  They are used to forward the parameter types of one
+        callable to another callable, a pattern commonly found in higher order
+        functions and decorators.  They are only valid when used in ``Concatenate``,
+        or s the first argument to ``Callable``. In Python 3.10 and higher,
+        they are also supported in user-defined Generics at runtime.
+        See class Generic for more information on generic types.  An
+        example for annotating a decorator::
+
+           T = TypeVar('T')
+           P = ParamSpec('P')
+
+           def add_logging(f: Callable[P, T]) -> Callable[P, T]:
+               '''A type-safe decorator to add logging to a function.'''
+               def inner(*args: P.args, **kwargs: P.kwargs) -> T:
+                   logging.info(f'{f.__name__} was called')
+                   return f(*args, **kwargs)
+               return inner
+
+           @add_logging
+           def add_two(x: float, y: float) -> float:
+               '''Add two numbers together.'''
+               return x + y
+
+        Parameter specification variables defined with covariant=True or
+        contravariant=True can be used to declare covariant or contravariant
+        generic types.  These keyword arguments are valid, but their actual semantics
+        are yet to be decided.  See PEP 612 for details.
+
+        Parameter specification variables can be introspected. e.g.:
+
+           P.__name__ == 'T'
+           P.__bound__ == None
+           P.__covariant__ == False
+           P.__contravariant__ == False
+
+        Note that only parameter specification variables defined in global scope can
+        be pickled.
+        """
+        args = object()
+        kwargs = object()
+
+        def __init__(self, name, *, bound=None, covariant=False, contravariant=False):
```

Comment:
```suggestion
        def __init__(self, name, *, bound=None, covariant=False, contravariant=False):
            super().__init__([self])
```
With this change, the runtime annotations of the `add_logging` example in the PEP look like this:
```
In [10]: add_logging.__annotations__
Out[10]: 
{'f': typing.Callable[[~P], ~R],
 'return': typing.Callable[[~P], typing.Awaitable[~R]]}
```
That's not exactly correct, but it's close enough that dynamic type checkers should be able to do something with it. With your current code, it produces `Callable[[], ~R]` because it iterates over the ParamSpec and finds nothing.

This feels a little bad because we're creating a list that contains itself, but I think it might not have any ill effects except for creating a reference cycle.

Reply:
Wow this is such a nice hack. The only gripe I have is that technically ``Callable[[~P], int]`` is invalid according to PEP 612 https://www.python.org/dev/peps/pep-0612/#id2 . And I really don't want to promote the invalid form even if its just showing it in repr, we already have enough improper usages of Callable as is. Hmm im not sure what route to take here 😔.

Reply:
That's true, but seeing `Callable[[], ~R]` in the repr (the current state) is pretty confusing too. My main motivation is to give dynamic type checkers something to work with here.

Reply:
You're right, either way both reprs will be wrong, so let's choose the one that actually benefits a bunch of people.